File: omnipod-dash/src/main/java/info/nightscout/androidaps/plugins/pump/omnipod/dash/driver/comm/scan/ScanCollector.java
package info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.scan;

import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import info.nightscout.androidaps.logging.AAPSLogger;
import info.nightscout.androidaps.logging.LTag;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.exceptions.DiscoveredInvalidPodException;
import info.nightscout.androidaps.plugins.pump.omnipod.dash.driver.comm.exceptions.ScanFailException;

public class ScanCollector extends ScanCallback {
    private final AAPSLogger logger;
    private final long podID;
    // there could be different threads calling the onScanResult callback
    private final ConcurrentHashMap<String, ScanResult> found;
    private int scanFailed;

    public ScanCollector(AAPSLogger logger, long podID) {
        this.podID = podID;
        this.logger = logger;
        this.found = new ConcurrentHashMap<String, ScanResult>();
    }

    @Override
    public void onScanResult(int callbackType, ScanResult result) {
        // callbackType will be ALL
        this.logger.debug(LTag.PUMPBTCOMM, "Scan found: "+result.toString());
        this.found.put(result.getDevice().getAddress(), result);
    }

    @Override
    public void onScanFailed(int errorCode) {
        this.scanFailed = errorCode;
        this.logger.warn(LTag.PUMPBTCOMM, "Scan failed with errorCode: "+errorCode);
        super.onScanFailed(errorCode);
    }

    public List<BleDiscoveredDevice> collect()
            throws ScanFailException {
        List<BleDiscoveredDevice> ret = new ArrayList<>();

        if (this.scanFailed != 0) {
            throw new ScanFailException(this.scanFailed);
        }

        logger.debug(LTag.PUMPBTCOMM, "ScanCollector looking for podID: " + this.podID);

        for (ScanResult result : this.found.values()) {
            try {
                BleDiscoveredDevice device = new BleDiscoveredDevice(result, this.podID);
                ret.add(device);
                logger.debug(LTag.PUMPBTCOMM, "ScanCollector found: " + result.toString() + "Pod ID: " + this.podID);
            } catch (DiscoveredInvalidPodException e) {
                logger.debug(LTag.PUMPBTCOMM, "ScanCollector: pod not matching" + e.toString());
                // this is not the POD we are looking for
            }
        }
        return Collections.unmodifiableList(ret);
    }
}

File: pump/omnipod/dash/src/main/kotlin/app/aaps/pump/omnipod/dash/driver/comm/scan/ScanCollector.kt
package app.aaps.pump.omnipod.dash.driver.comm.scan

import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import app.aaps.core.interfaces.logging.AAPSLogger
import app.aaps.core.interfaces.logging.LTag
import app.aaps.pump.omnipod.dash.driver.comm.exceptions.ScanException
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap

class ScanCollector(private val logger: AAPSLogger, private val podID: Long) : ScanCallback() {

    // there could be different threads calling the onScanResult callback
    private val found: ConcurrentHashMap<String, ScanResult> = ConcurrentHashMap()
    private var scanFailed = 0
    override fun onScanResult(callbackType: Int, result: ScanResult) {
        // callbackType will be ALL
        logger.debug(LTag.PUMPBTCOMM, "Scan found: $result")
        found[result.device.address] = result
    }

    override fun onScanFailed(errorCode: Int) {
        logger.warn(LTag.PUMPBTCOMM, "Scan failed with errorCode: $errorCode")
        super.onScanFailed(errorCode)
    }

    @Throws(ScanException::class) fun collect(): List<BleDiscoveredDevice> {
        val ret: MutableList<BleDiscoveredDevice> = ArrayList()
        if (scanFailed != 0) {
            throw ScanException(scanFailed)
        }
        logger.debug(LTag.PUMPBTCOMM, "ScanCollector looking for podID: $podID")
        for (result in found.values) {
            try {
                result.scanRecord?.let {
                    val device = BleDiscoveredDevice(result, it, podID)
                    ret.add(device)
                    logger.debug(LTag.PUMPBTCOMM, "ScanCollector found: " + result.toString() + "Pod ID: " + podID)
                }
            } catch (e: DiscoveredInvalidPodException) {
                logger.debug(LTag.PUMPBTCOMM, "ScanCollector: pod not matching$e")
                // this is not the POD we are looking for
            }
        }
        return Collections.unmodifiableList(ret)
    }
}
